From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Thu, 23 Feb 2023 21:08:30 +0100
Subject: [PATCH] All nether bricks count as fortress for mob spawning

License: MIT (https://opensource.org/licenses/MIT)
Suki - https://github.com/SuCraft/Suki

This patch is based on the following patch:
"Do not require nether fortress structure for mob spawning"
By: chickeneer <emcchickeneer@gmail.com>
As part of: EmpireCraft (https://github.com/starlis/empirecraft)
Licensed under: MIT (https://opensource.org/licenses/MIT)

* EmpireCraft description *

Due to failed fortress conversions - some structure data is missing.
This allows players to make their own fortress with nether bricks

diff --git a/src/main/java/net/minecraft/world/level/NaturalSpawner.java b/src/main/java/net/minecraft/world/level/NaturalSpawner.java
index 969933ff94a1ae5a36d0550cd10d9e9803813136..f9a5068353374ab29df2e7066a0d0dc2cc5ce75a 100644
--- a/src/main/java/net/minecraft/world/level/NaturalSpawner.java
+++ b/src/main/java/net/minecraft/world/level/NaturalSpawner.java
@@ -48,6 +48,7 @@ import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.level.pathfinder.PathComputationType;
 import net.minecraft.world.level.storage.LevelData;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.World;
 import org.slf4j.Logger;
 import org.bukkit.craftbukkit.util.CraftSpawnCategory;
 import org.bukkit.entity.SpawnCategory;
@@ -407,6 +408,11 @@ public final class NaturalSpawner {
 
     public static boolean isInNetherFortressBounds(BlockPos pos, ServerLevel world, MobCategory spawnGroup, StructureManager structureAccessor) {
         if (spawnGroup == MobCategory.MONSTER && world.getBlockState(pos.below()).is(Blocks.NETHER_BRICKS)) {
+            // Suki start - all nether bricks count as fortress
+            if (world.sukiConfig().gameplayMechanics.mobSpawning.allNetherBricksCountAsFortress && world.getWorld().getEnvironment() == World.Environment.NETHER) {
+                return true;
+            }
+            // Suki end - all nether bricks count as fortress
             Structure structure = (Structure) structureAccessor.registryAccess().registryOrThrow(Registries.STRUCTURE).get(BuiltinStructures.FORTRESS);
 
             return structure == null ? false : structureAccessor.getStructureAt(pos, structure).isValid();
diff --git a/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java b/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java
index 92ee3afd796692599df07631bc57443d14979525..5b4c217bc30f9ac0f2f6584f391299a804ffc8a8 100644
--- a/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java
+++ b/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java
@@ -53,6 +53,13 @@ public class SukiWorldConfiguration extends ConfigurationPart {
         }
         // Suki end - never despawn entities due to distance
 
+        // Suki start - all nether bricks count as fortress
+        public MobSpawning mobSpawning;
+        public class MobSpawning extends ConfigurationPart {
+            public boolean allNetherBricksCountAsFortress = false;
+        }
+        // Suki end - all nether bricks count as fortress
+
     }
 
     public Misc misc;
